packages/cli/generators/repository/index.js

```diff
@@ -0,0 +1,393 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/cli
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+'use strict';
+const _ = require('lodash');
+const ArtifactGenerator = require('../../lib/artifact-generator');
+const debug = require('../../lib/debug')('repository-generator');
+const inspect = require('util').inspect;
+const path = require('path');
+const chalk = require('chalk');
+const utils = require('../../lib/utils');
+const connectors = require('../datasource/connectors.json');
+const tsquery = require('@phenomnomnominal/tsquery').tsquery;
+
+const VALID_CONNECTORS_FOR_REPOSITORY = ['KeyValueModel', 'PersistedModel'];
+const KEY_VALUE_CONNECTOR = ['KeyValueModel'];
+
+const DEFAULT_CRUD_REPOSITORY = 'DefaultCrudRepository';
+const KEY_VALUE_REPOSITORY = 'KeyValueRepository';
+
+const REPOSITORY_KV_TEMPLATE = 'repository-kv-template.ts.ejs';
+const REPOSITORY_CRUD_TEMPLATE = 'repository-crud-default-template.ts.ejs';
+
+const VALID_PROPERTY_ID_TYPES = ['number', 'string', 'object'];
+
+const PROMPT_MESSAGE_MODEL =
+  'Select the model(s) you want to generate a repository';
+const PROMPT_MESSAGE_DATA_SOURCE = 'Please select the datasource';
+const ERROR_READING_FILE = 'Error reading file';
+const ERROR_NO_DATA_SOURCES_FOUND = 'No datasources found in';
+const ERROR_NO_MODELS_FOUND = 'No models found in';
+const ERROR_NO_MODEL_SELECTED = 'You did not select any model';
+
+module.exports = class RepositoryGenerator extends ArtifactGenerator {
+  // Note: arguments and options should be defined in the constructor.
+  constructor(args, opts) {
+    super(args, opts);
+  }
+
+  async _getModelFields(modelName) {
+    let retArray = [];
+    let fileContent = '';
+
+    let modelFile = path.join(
+      this.artifactInfo.modelDir,
+      modelName.toLowerCase() + '.model.ts',
+    );
+
+    try {
+      fileContent = this.fs.read(modelFile, {});
+    } catch (err) {
+      debug(`${ERROR_READING_FILE} ${modelFile}: ${err.message}`);
+      return this.exit(err);
+    }
+
+    const ast = tsquery.ast(fileContent);
+    const nodes = tsquery(ast, ':matches([kindName="PropertyDeclaration"])');
+
+    for (let x of nodes) {
+      retArray.push(x.name.escapedText);
+    }
+
+    return retArray;
+  }
+
+  async _getModelIdType(modelName) {
```

Comment:
Could you please move the code detecting the "id" property into as standalone helper file that can be used by other generators? It will also make it easier to test this functionality.

A few non-obvious test cases that come to my mind:

The id property is defined via `@model` definition:

```ts
@model({
  name: 'Product',
  properties: {
    id: {type: 'number', id: true}
    name: {type: 'string', required: true},
  },
})
class Product extends Entity {
  id: number;
  name: string;
}
```

The model does not use any decorator:

```ts
class Product extends Entity {
  static definition = {
    name: 'Product',
    properties: {
      id: {type: 'number', id: true}
      name: {type: 'string', required: true},
    },
  };

  id: number;
  name: string;
}
```

Reply:
Yes, that's true, there are many variations. One that is a little bit more complex is when you import the JSON definition file from somewhere else (ie: from `@yourcompany/models`) . I believe we should tackle the most common scenarios and then fine tune it to land the PR?

Reply:
>  I believe we should tackle the most common scenarios and then fine tune it to land the PR?

Yes please!

Just make sure there is a workaround solution for people using an uncommon scenario. I think your current implementation, where the user can enter a custom value if the AST-based reader did not find anything, is a great solution for that 👍